src/languageclient/startArmLanguageServer.ts

```diff
@@ -1,53 +1,54 @@
-// ----------------------------------------------------------------------------
-// Copyright (c) Microsoft Corporation.  All rights reserved.
-// ----------------------------------------------------------------------------
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
 
-import * as fs from 'fs';
-import * as os from 'os';
+import * as fse from 'fs-extra';
 import * as path from 'path';
 import { ExtensionContext, workspace } from 'vscode';
-import { callWithTelemetryAndErrorHandlingSync, parseError, TelemetryProperties } from 'vscode-azureextensionui';
-import { Message } from 'vscode-jsonrpc';
-import { CloseAction, ErrorAction, ErrorHandler, LanguageClient, LanguageClientOptions, ServerOptions } from 'vscode-languageclient';
+import { callWithTelemetryAndErrorHandling, callWithTelemetryAndErrorHandlingSync, IActionContext, parseError } from 'vscode-azureextensionui';
+import { LanguageClient, LanguageClientOptions, ServerOptions } from 'vscode-languageclient';
+import { dotnetAcquire, ensureDotnetDependencies } from '../acquisition/dotnetAcquisition';
 import { armDeploymentLanguageId } from '../constants';
 import { ext } from '../extensionVariables';
 import { armDeploymentDocumentSelector } from '../supported';
+import { WrappedErrorHandler } from './WrappedErrorHandler';
 
-const languageServerName = 'ARM Language Server';
-const languageServerFolderName = 'LanguageServerBin';
+export const languageServerName = 'ARM Language Server';
+const languageServerFolderName = 'languageServer';
 const languageServerDllName = 'Microsoft.ArmLanguageServer.dll';
-let serverStartMs: number;
-const languageServerErrorTelemId = "Language Server Error";
+export let serverStartMs: number;
+export const languageServerErrorTelemId = 'Language Server Error';
+const defaultTraceLevel = 'Warning';
+const dotnetVersion = '2.2';
+
+export enum LanguageServerState {
+    NotStarted,
+    Starting,
+    Failed,
+    Started,
+}
+export let languageServerState: LanguageServerState = LanguageServerState.NotStarted;
 
-export function startArmLanguageServer(context: ExtensionContext): void {
-    callWithTelemetryAndErrorHandlingSync('startArmLanguageClient', () => {
+export async function startArmLanguageServer(): Promise<void> {
+    languageServerState = LanguageServerState.Starting;
+    try {
         // The server is implemented in .NET Core. We run it by calling 'dotnet' with the dll as an argument
+        let serverDllPath: string = findLanguageServer(ext.context);
+        let dotnetExePath: string = await acquireDotnet(serverDllPath);
+        await ensureDependencies(dotnetExePath, serverDllPath);
```

Comment:
So note: This will attempt to boot your server with 0 arguments. If there's a piece of your server that you want to try out (any complex dependencies, doubt you have native ones) I'd be sure that call path is hit during this boot to ensure that those assemblies can be loaded/executed properly.

Reply:
Hmm, good point.  I don't think we ever actually hit needing additional dependencies, though. I haven't seen it run yet.

I guess I'd need to add a hidden command-line argument to somehow hit the code analysis logic and then quit, since the server never exits normally.

Reply:
I'm not actually clear on what dependencies can be missing, do you know?

Reply:
Thanks!